### PR TITLE
Addition of Peppol e-invoice xml 

### PR DESCRIPTION
**Pull request by Ninarchive**
First version of peppol xml generation

Note that the xml can't be saved in the database yet. A new field has to be added to Attachment to be able to save it together with the pdf.

### CODE REVIEW

File: backend/src/controllers/utils/index.ts [hunk 1]
```diff
@@ -72,3 +81,186 @@ export const getTemplatesPath = (): string => {
...
+
+    cityRef = companyConfig.company.city.trim();
+    const supplierPostalAddress = new PostalAddress({
+      streetName: companyConfig.company.address,
+      cityName: cityRef,
+      country: new Country({ identificationCode: DEFAULT_COUNTRY_CODE })
+    });
+    const supplierLegalEntity = new PartyLegalEntity({
+      registrationName: companyConfig.company.name,
+      companyID: companyConfig.company.btw
```
Comment (Laoujin) [inline backend/src/controllers/utils/index.ts:125]: Are we sure that the CompanyID is the VAT nr?

File: backend/src/controllers/utils/index.ts [hunk 2]
```diff
@@ -72,3 +81,186 @@ export const getTemplatesPath = (): string => {
...
+    const customerEndpointScheme = ENDPOINT_SCHEMES.find(scheme => scheme.country === savedInvoice.client.country);
+    const customerEndpointID = new UdtIdentifier(savedInvoice.client.btw, {
+      schemeID: customerEndpointScheme ? customerEndpointScheme.schemeID : ''
+    });
+    const customerParty = new Party({
+      EndpointID: customerEndpointID,
+      partyLegalEntities: [customerLegalEntity],
+      postalAddress: customerPostalAddress,
+      partyTaxSchemes: [new PartyTaxScheme({
+        companyID: savedInvoice.client.btw,
```
Comment (Laoujin) [inline backend/src/controllers/utils/index.ts:164]: Same here: companyID is the VAT nr?

File: backend/src/controllers/utils/index.ts [hunk 3]
```diff
@@ -72,3 +81,186 @@ export const getTemplatesPath = (): string => {
...
+        percent: '0',
+        taxScheme: taxSchemeIDVAT
+      }
+    } else {
+      taxObject = {
+        id: 'S',
+        percent: '21',
+        taxScheme: taxSchemeIDVAT
+      }
+      classifiedTaxObject = taxObject;
```
Comment (Laoujin) [inline backend/src/controllers/utils/index.ts:200]: Add some constants explaining what AE, S and VATEX-EU-AE is

File: backend/src/controllers/utils/index.ts [hunk 4]
```diff
@@ -72,3 +81,186 @@ export const getTemplatesPath = (): string => {
...
+    const orderRef = savedInvoice.orderNr && savedInvoice.orderNr.trim().length !== 0 ? savedInvoice.orderNr : savedInvoice._id;
+
+    invoiceXml.addProperty('xmlns', 'urn:oasis:names:specification:ubl:schema:xsd:Invoice-2');
+    invoiceXml.addProperty('xmlns:cac', 'urn:oasis:names:specification:ubl:schema:xsd:CommonAggregateComponents-2');
+    invoiceXml.addProperty('xmlns:cbc', 'urn:oasis:names:specification:ubl:schema:xsd:CommonBasicComponents-2');
+    invoiceXml.setCustomizationID('urn:cen.eu:en16931:2017#compliant#urn:fdc:peppol.eu:2017:poacc:billing:3.0');
+    invoiceXml.setProfileID('urn:fdc:peppol.eu:2017:poacc:billing:01:1.0');
+    invoiceXml.setIssueDate(moment(savedInvoice.date).format('YYYY-MM-DD'));
+    invoiceXml.setDueDate(moment(savedInvoice.date).add(30, 'days').format('YYYY-MM-DD'));
+    invoiceXml.setInvoiceTypeCode('380');
```
Comment (Laoujin) [inline backend/src/controllers/utils/index.ts:232]: Put the 380 in a constant
For these constants, maybe you could even add a little JSDoc with the link to further details (the webpages you showed me)

```js
/** URL TO DOCS */
const NormalInvoice = '380';
```

(or something)

File: backend/src/controllers/utils/index.ts [hunk 5]
```diff
@@ -72,3 +81,186 @@ export const getTemplatesPath = (): string => {
...
+    invoiceXml.setID(savedInvoice.number.toString());
+    invoiceXml.setOrderReference(new OrderReference({ id: orderRef }));
+    invoiceXml.addTaxTotal(taxTotal);
+    invoiceXml.addPaymentMeans(paymentMeans);
+
+    savedInvoice.lines.forEach((line, index) => {
+      const unitCode = UNIT_CODES.find(unitCode => unitCode.unit === line.type);
+      const invoiceLine = new InvoiceLine({
+        id: (index + 1).toString(),
+        invoicedQuantity: new UdtQuantity(line.amount.toString(), { unitCode: unitCode ? unitCode.code : 'C64' }),
```
Comment (Laoujin) [inline backend/src/controllers/utils/index.ts:246]: Constant for C64

File: backend/src/models/invoices.ts [hunk 6]
```diff
@@ -75,3 +75,27 @@ export const INVOICE_EXCEL_HEADERS = [
...
+  { country: 'GB', schemeID: '9932' }
+];
+
+export const UNIT_CODES = [
+  { unit: 'daily', code: 'DAY' },
+  { unit: 'hourly', code: 'HUR' },
+  { unit: 'km', code: 'KMT' },
+  { unit: 'items', code: 'NAR' },
+  { unit: 'other', code: 'C64' }
+];
```
Comment (Laoujin) [inline backend/src/models/invoices.ts:101]: Can these `'hourly'` etc reference to the existing type?

```
type UnitCodes = { unit: EditClientRateType, code: string }
export const UNIT_CODES: UnitCodes[] = [
```

File: backend/src/controllers/utils/index.ts [hunk 7]
```diff
@@ -72,3 +81,186 @@ export const getTemplatesPath = (): string => {
...
+    const supplierEndpointID = new UdtIdentifier(companyConfig.company.btw, {
+      schemeID: supplierEndpointScheme ? supplierEndpointScheme.schemeID : '9925'
+    });
+
+    const supplierParty = new Party({
+      EndpointID: supplierEndpointID,
+      partyLegalEntities: [supplierLegalEntity],
+      postalAddress: supplierPostalAddress,
+      contact: new Contact({
+        name: companyConfig.company.website,
```
[the remark this answers is not in the document]
Comment (Ninarchive) [reply backend/src/controllers/utils/index.ts:138]: Our website is mentioned on the pdf, so I tried to find a place to add it to the xml too. This seemed the most appriopate field available, there is no specific element to add website information. And the example invoice I found also placed a website reference in the name field, so I don't think it's that out of place :)

File: backend/src/controllers/config.ts [hunk 8]
```diff
@@ -1,15 +1,15 @@
...
-import {ICompanyConfig} from '../models/config';
-import {getTemplatesPath} from './utils';
-import {CollectionNames, updateAudit} from '../models/common';
-import {ConfacRequest} from '../models/technical';
-import {saveAudit} from './utils/audit-logs';
+import { ICompanyConfig } from '../models/config';
+import { getTemplatesPath } from './utils';
+import { CollectionNames, updateAudit } from '../models/common';
+import { ConfacRequest } from '../models/technical';
+import { saveAudit } from './utils/audit-logs';
```
[the remark this answers is not in the document]
Comment (Ninarchive) [reply backend/src/controllers/config.ts]: Fixed

File: backend/src/controllers/invoices.ts [hunk 9]
```diff
@@ -252,3 +266,33 @@ export const generateExcelForInvoicesController = async (req: Request, res: Resp
 
   return res.send(excel);
 };
+
+
+export const getInvoiceXmlController = async (req: Request, res: Response) => {
+  const { id: invoiceId }: { id: string; } = req.body;
+  const invoiceAttachments: IAttachmentCollection | null = await req.db.collection(CollectionNames.ATTACHMENTS)
+    .findOne({ _id: new ObjectID(invoiceId) as ObjectID });
```
[the remark this answers is not in the document]
Comment (Ninarchive) [reply backend/src/controllers/invoices.ts]: Changed

File: backend/src/controllers/invoices.ts [hunk 10]
```diff
@@ -252,3 +266,33 @@ export const generateExcelForInvoicesController = async (req: Request, res: Resp
...
+  if (invoiceAttachments && invoiceAttachments.xml) {
+    return res.type('application/xml').send(atob(invoiceAttachments.xml.toString()));
+  } else {
+    return res.status(500).send('No xml found');
+  }
+};
+
+export const testXMlController = async (req: Request, res: Response) => {
```
[the remark this answers is not in the document]
Comment (Ninarchive) [reply backend/src/controllers/invoices.ts]: Done

File: backend/src/controllers/utils/index.ts [hunk 11]
```diff
@@ -72,3 +81,186 @@ export const getTemplatesPath = (): string => {
   }
   return './templates/';
 };
+
+
+export const createXml = (savedInvoice: IInvoice, companyConfig: ICompanyConfig): string => {
```
[the remark this answers is not in the document]
Comment (Ninarchive) [reply backend/src/controllers/utils/index.ts]: Changed

File: backend/src/controllers/utils/index.ts [hunk 12]
```diff
@@ -72,3 +81,186 @@ export const getTemplatesPath = (): string => {
...
+      country: new Country({ identificationCode: DEFAULT_COUNTRY_CODE })
+    });
+    const supplierLegalEntity = new PartyLegalEntity({
+      registrationName: companyConfig.company.name,
+      companyID: companyConfig.company.btw
+    });
+
+    const supplierEndpointScheme = ENDPOINT_SCHEMES.find(scheme => scheme.country === DEFAULT_COUNTRY_CODE);
+    const supplierEndpointID = new UdtIdentifier(companyConfig.company.btw, {
+      schemeID: supplierEndpointScheme ? supplierEndpointScheme.schemeID : '9925'
```
[the remark this answers is not in the document]
Comment (Ninarchive) [reply backend/src/controllers/utils/index.ts]: Done

File: backend/src/controllers/utils/index.ts (hunk 3 above)
Comment (Ninarchive) [reply backend/src/controllers/utils/index.ts:206]: Done

File: backend/src/controllers/utils/index.ts [hunk 13]
```diff
@@ -72,3 +81,186 @@ export const getTemplatesPath = (): string => {
...
+      taxAmount: new UdtAmount(savedInvoice.money.totalTax.toFixed(2), currencyID),
+      taxSubtotals: [new TaxSubtotal({
+        taxableAmount: new UdtAmount(savedInvoice.money.totalWithoutTax.toFixed(2), currencyID),
+        taxAmount: new UdtAmount(savedInvoice.money.totalTax.toFixed(2), currencyID),
+        taxCategory: new TaxCategory(taxObject)
+      })]
+    });
+
+    const paymentMeans = new PaymentMeans({
+      paymentMeansCode: '31',
```
[the remark this answers is not in the document]
Comment (Ninarchive) [reply backend/src/controllers/utils/index.ts]: Done

File: backend/src/controllers/utils/index.ts (hunk 5 above)
Comment (Ninarchive) [reply backend/src/controllers/utils/index.ts]: Done

File: backend/src/models/invoices.ts [hunk 14]
```diff
@@ -75,3 +75,27 @@ export const INVOICE_EXCEL_HEADERS = [
   'Without Tax', 'Tax', 'Total', 'Verified', 'Discount',
   'First line desc', 'Id',
 ];
+
+export const COUNTRY_CODES = [
```
[the remark this answers is not in the document]
Comment (Ninarchive) [reply backend/src/models/invoices.ts:85]: Done

File: backend/src/models/invoices.ts (hunk 6 above)
Comment (Ninarchive) [reply backend/src/models/invoices.ts:116]: Done

File: backend/src/controllers/utils/index.ts [hunk 15]
```diff
@@ -72,3 +80,204 @@ export const getTemplatesPath = (): string => {
...
+    invoiceXml.setID(savedInvoice.number.toString());
+    invoiceXml.setOrderReference(new OrderReference({id: orderRef}));
+    invoiceXml.addTaxTotal(taxTotal);
+    invoiceXml.addPaymentMeans(paymentMeans);
+
+    /** more info on unit codes: https://docs.peppol.eu/poacc/billing/3.0/codelist/UNECERec20/ 
+     * code C62 is a general code meaning 'one' or 'unit'
+    */
+    const defaultUnitCode = 'C62';
+    savedInvoice.lines.forEach((line, index) => {
```
Comment (Laoujin) [inline backend/src/controllers/utils/index.ts:259]: fyi (for your information) -- you can leave this code AS IS

```
savedInvoice.lines.forEach(...)
```

Avoid this kind of construct (unless good reason to do so). A forEach means that there is going to be one or more side effects in the function. You need to read the entire function and hope that you spot all the data-manipulations. This makes it difficult to understand what is going on / going to happen.

Prefer using side-effect free code:
```
const xmlInvoiceLines = savedInvoice.lines.map((line, index) => { ... return new InvoiceLine(...); }
invoiceXml.addInvoiceLines(xmlInvoiceLines)
```

It looks almost the same, but it makes your functions pure and side-effects evident...

File: backend/src/controllers/utils/index.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (Laoujin) [reply backend/src/controllers/utils/index.ts]: Ok, if the reference site also did it that way, it should be good :)

File: backend/src/controllers/utils/index.ts (hunk 1 above)
Comment (Laoujin) [reply backend/src/controllers/utils/index.ts]: CBE == BCE == KBO (en in company model == company.btw)

CBE is the english translation of KBO

File: backend/src/controllers/utils/index.ts (hunk 15 above)
Comment (Ninarchive) [reply backend/src/controllers/utils/index.ts:259]: Thanks for the feedback, I hadn't thought about this! I will keep this in mind in the future

**Pull request merged**